encoding/dictionary.go

```diff
@@ -0,0 +1,96 @@
+package encoding
+
+import (
+	"encoding/binary"
+	"io"
+
+	"github.com/xichen2020/eventdb/generated/proto/encodingpb"
+)
+
+type dictionaryStringIterator struct {
+	mapping   []string
+	meta      encodingpb.StringMeta
+	reader    io.Reader
+	container []byte
+	curr      string
+	err       error
+	closed    bool
+}
+
+func newDictionaryStringIterator(reader io.Reader, meta encodingpb.StringMeta, mapping []string) ForwardStringIterator {
+	return &dictionaryStringIterator{
+		reader:    reader,
+		meta:      meta,
+		mapping:   mapping,
+		container: make([]byte, 8),
+	}
+}
+
+func (d *dictionaryStringIterator) Next() bool {
+	if d.closed {
+		return false
+	}
+
+	_, d.err = d.reader.Read(d.container[:d.meta.BytesPerRecord])
+	if d.err != nil {
+		return false
+	}
+	idx := int(binary.LittleEndian.Uint64(d.container))
+
+	d.curr = d.mapping[idx]
+
+	return true
+}
+
+func (d *dictionaryStringIterator) Current() string {
+	return d.curr
+}
+
+func (d *dictionaryStringIterator) Err() error {
+	return d.err
+}
+
+// Reset of internal data not supported for dictionary iterator.
+func (d *dictionaryStringIterator) Reset(values []string) {}
+
+func (d *dictionaryStringIterator) Close() error {
+	d.closed = true
+	return nil
+}
+
+// Dictionary encoding strategy is to write
+// and then the bytes for the string.
+func (enc *StringEnc) encodeDictionary(writer io.Writer, meta encodingpb.StringMeta, values RewindableStringIterator, mapping map[string]int64, reverseMapping []string) error {
+	// Write out the reverseMapping first.
+	if err := ProtoEncode(&encodingpb.StringMapping{
+		Data: reverseMapping,
+	}, writer); err != nil {
+		return err
+	}
+
+	// Write out the dictionary values.
+	for values.Next() {
+		idx := mapping[values.Current()]
+		binary.LittleEndian.PutUint64(enc.container[:8], uint64(idx))
+		if _, err := writer.Write(enc.container[:meta.BytesPerRecord]); err != nil {
+			return err
+		}
+	}
+	return values.Err()
+}
+
+func (dec *StringDec) decodeDictionary(reader io.Reader, meta encodingpb.StringMeta) (ForwardStringIterator, error) {
```

Comment:
👍

Reply:
Hm I'd prefer to move this logic to the iterator since it logically belongs to the `dictionaryIterator` which presumably needs to decode the dictionary.

Reply:
Ah okay, it'll be a little ugly but I'll shove the decode logic into the iterator.